profiles/common/modules/features/nexteuropa_dgt_connector/tmgmt_dgt_connector/plugins/tmgmt_dgt_connector.plugin.inc

```diff
@@ -0,0 +1,284 @@
+<?php
+
+/**
+ * @file
+ * Provides TMGMT DGT Connector translation plugin controller.
+ */
+
+use EC\Poetry\Poetry;
+use EC\Poetry\Messages\Components;
+use EC\Poetry\Exceptions;
+
+/**
+ * DGT Connector translation plugin controller.
+ *
+ * It's now mostly based on TMGMTPoetryTranslatorPluginController
+ * Later this should extend TMGMTDefaultTranslatorPluginController and
+ * TMGMTPoetryTranslatorPluginController should be deprecated.
+ */
+class TmgmtDgtConnectorTranslatorPluginController extends TMGMTPoetryTranslatorPluginController {
+
+  /**
+   * Implements TMGMTTranslatorPluginControllerInterface::canTranslation().
+   *
+   * We will check are source and target language supported
+   * and the number of items.
+   */
+  public function canTranslate(TMGMTTranslator $translator, TMGMTJob $job) {
+    // Check that the item does not have pending requests.
+    $items = $job->getItems();
+    $item = array_shift($items);
+    if (!empty($item)) {
+      $results = db_select('tmgmt_job_item', 'tji')
+        ->fields('tji', array('tjid', 'tjiid'))
+        ->condition('item_type', $item->item_type, '=')
+        ->condition('item_id', $item->item_id, '=')
+        ->execute()
+        ->fetchAllAssoc('tjid');
+      if ($results) {
+        foreach ($results as $result) {
+          if ($job = tmgmt_job_load($result->tjid)) {
+            if ($job->hasPoetryTranslator() && $job->getState() == 1) {
+              // Check the poetry status.
+              $result_ps = db_select('poetry_status', 'ps')
+                ->fields('ps', array('status'))
+                ->condition('tjiid', $result->tjiid, '=')
+                ->execute()
+                ->fetchAssoc();
+              if (empty($result_ps) || (!empty($result_ps) && $result_ps['status'] == POETRY_STATUS_MESSAGE_PEN)) {
+                return FALSE;
+              }
+            }
+          }
+        }
+      }
+    }
+    // All languages are supported at the moment.
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getNotCanTranslateReason(TMGMTJob $job) {
+    return t('An active translation for this item already exists');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function requestTranslation(TMGMTJob $job) {
+
+    // Request translation from Poetry.
+    $response = $this->sendRequest($job);
+
+    // Handling response.
+    $response_xml = simplexml_load_string($response);
+
+    // If it is not an xml response, the SOAP request failed.
+    if (!$response_xml) {
```

Comment:
I think the try/catch wrapping the use of the Poetry component is in the right place, to catch exactly the errors coming from the component.

This method is "cloned" from TMGMTPoetryTranslatorPluginController. I'm using this strategy to not break functionality. And I'm adapting and refactoring piece by piece.

I agree, we have a large room for improvement here. But the response from Poetry component is still a WIP so I didn't put much attention in this part of the code. I pledge it will be improved later.